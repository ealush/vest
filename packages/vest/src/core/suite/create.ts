import asArray from 'asArray';
import genId from 'genId';
import isFunction from 'isFunction';
import throwError from 'throwError';
import createState from 'vast';

import createStateRef from 'createStateRef';
import context from 'ctx';
import { IVestResult, produceFullResult } from 'produce';
import { produceDraft, TDraftResult } from 'produceDraft';
import {
  usePending,
  useTestObjects,
  useCarryOverTests,
  useLagging,
} from 'stateHooks';


export default function create<T extends (...args: any[]) => void>(
  suiteCallback: T
): {
  (...args: Parameters<T>): IVestResult;

  get: () => TDraftResult;
  reset: () => void;
  remove: (fieldName: string) => void;
  subscribe: (handler: () => void) => void;
} {
  if (!isFunction(suiteCallback)) {
    throwError(
      'Suite initialization error. Expected `tests` to be a function.'
    );
  }

  const handlers: ((...args: unknown[]) => void)[] = [];
  const state = createState(() => {
    handlers.forEach(fn =>
      fn({
        suiteState: stateRef,
        type: 'suiteStateUpdate',
      })
    );
  });

  const stateRef = createStateRef(state, { suiteId: genId() });

  interface IVestSuite {
    (...args: Parameters<T>): IVestResult;

    get: () => TDraftResult;
    reset: () => void;
    remove: (fieldName: string) => void;
    subscribe: (handler: () => void) => void;
  }

  const suite: IVestSuite = Object.assign(
    context.bind({ stateRef }, (...args: unknown[]) => {
      const [previousTestObjects] = useTestObjects();
      const [, setCarryOverTests] = useCarryOverTests();
      const [pending] = usePending();
      const [prevLagging, setLagging] = useLagging();
      state.reset();
      setCarryOverTests(() => previousTestObjects);

      // Move all the active pending tests to the lagging array
      setLagging(pending.concat(prevLagging));

      // Run the consumer's callback
      suiteCallback(...args);

      const res = produceFullResult();

      return res;
    }),
    {
      get: context.bind({ stateRef }, produceDraft),
      remove: context.bind({ stateRef }, name => {
        const [testObjects] = useTestObjects();

        // We're mutating the array in `cancel`, so we have to first copy it.
        asArray(testObjects).forEach(testObject => {
          if (testObject.fieldName === name) {
            testObject.cancel();
          }
        });
      }),
      reset: state.reset,
      subscribe(
        handler: (stateEvent: {
          type: string;
          suiteState: typeof stateRef;
        }) => void
      ) {
        if (!isFunction(handler)) return;

        handlers.push(handler);

        handler({
          type: 'suiteSubscribeInit',
          suiteState: stateRef,
        });
      },
    }
  );

  return suite;
}
