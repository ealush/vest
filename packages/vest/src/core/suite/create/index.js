import genId from '../../../lib/id';
import throwError from '../../../lib/throwError';
import context from '../../context';
import produce from '../../produce';
import useTestCallbacks from '../../produce/useTestCallbacks';
import state from '../../state';
import mergeExcludedTests from '../../test/lib/mergeExcludedTests';
import usePending from '../../test/lib/pending/usePending';
import useTestObjects from '../../test/useTestObjects';
import useSuiteId from '../useSuiteId';
/**
 * Initializes a validation suite, creates a validation context.
 * @param {String} name     Identifier for validation suite.
 * @param {Function} tests  Validation suite body.
 * @returns {Function} validator function.
 */
const createSuite = (...args) => {
  const [tests, name] = args.reverse();

  if (typeof tests !== 'function') {
    throwError(
      'Suite initialization error. Expected `tests` to be a function.'
    );
  }

  const stateRef = state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [genId(), name]],
    useTestCallbacks,
    useTestObjects,
  });

  return Object.defineProperties(
    context.bind({ stateRef }, (...args) => {
      stateRef.unshift();

      tests.apply(null, args);
      mergeExcludedTests();

      return produce();
    }),
    {
      get: {
        value: context.bind({ stateRef }, produce, { draft: true }),
      },
      name: { value: 'validate' },
      reset: { value: stateRef.reset },
    }
  );
};

export default createSuite;
