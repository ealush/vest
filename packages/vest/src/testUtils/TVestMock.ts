import { TFieldName, TGroupName } from 'SuiteResultTypes';
import * as vest from 'vest';

export type TVestMock = typeof vest;

export type TTestSuiteCallback = (...args: any[]) => void;
export type TTestSuite = vest.Suite<TFieldName, TGroupName, TTestSuiteCallback>;
