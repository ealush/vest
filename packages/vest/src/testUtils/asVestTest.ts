import { invariant } from 'vest-utils';

import { ErrorStrings } from 'ErrorStrings';
import { TIsolateTest } from 'IsolateTest';
import { isIsolateTest } from 'isIsolateTest';

export function asVestTest(value: any): TIsolateTest {
  invariant(isIsolateTest(value), ErrorStrings.EXPECTED_VEST_TEST);
  return value;
}
