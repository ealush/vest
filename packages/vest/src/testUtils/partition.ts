export default function partition<T>(
  array: T[],
  // eslint-disable-next-line no-unused-vars
  predicate: (value: T, index: number, array: T[]) => boolean
): [T[], T[]] {
  return array.reduce(
    (partitions: [T[], T[]], value, number) => {
      partitions[predicate(value, number, array) ? 0 : 1].push(value);
      return partitions;
    },
    [[], []]
  );
}
