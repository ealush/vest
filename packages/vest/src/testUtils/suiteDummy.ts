import { OneOrMoreOf, asArray, Maybe } from 'vest-utils';

import { dummyTest } from './testDummy';

import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { optional, create, skip, SuiteResult } from 'vest';

export function failing(failingFields?: OneOrMoreOf<string>) {
  return createSuiteRunResult(failingFields, fieldName => {
    dummyTest.failing(fieldName);
  });
}

export function warning(failingFields?: OneOrMoreOf<string>) {
  return createSuiteRunResult(failingFields, fieldName => {
    dummyTest.failingWarning(fieldName);
  });
}

export function failingAsync(failingFields?: OneOrMoreOf<string>) {
  return createSuiteRunResult(failingFields, fieldName => {
    dummyTest.failingAsync(fieldName);
  });
}

export function passing(fields?: OneOrMoreOf<string>) {
  return createSuiteRunResult(fields, fieldName => {
    dummyTest.passing(fieldName);
  });
}

export function passingWithUntestedOptional(
  optionals: OneOrMoreOf<string> = 'optional_field',
  required: OneOrMoreOf<string> = 'field_1'
) {
  return create(() => {
    optional(optionals);
    skip(optionals);

    asArray(optionals).forEach(fieldName => {
      dummyTest.failing(fieldName);
    });

    asArray(required).forEach(fieldName => {
      dummyTest.passing(fieldName);
    });
  })();
}

export function passingWithOptional(
  optionals: OneOrMoreOf<string> = 'optional_field',
  required: OneOrMoreOf<string> = 'field_1'
) {
  return create(() => {
    optional(optionals);

    asArray(optionals).forEach(fieldName => {
      dummyTest.passing(fieldName);
    });

    asArray(required).forEach(fieldName => {
      dummyTest.passing(fieldName);
    });
  })();
}

export function failingOptional(
  optionals: OneOrMoreOf<string> = 'optional_field',
  required: OneOrMoreOf<string> = 'field_1'
) {
  return create(() => {
    optional(optionals);

    asArray(optionals).forEach(fieldName => {
      dummyTest.failing(fieldName);
    });

    asArray(required).forEach(fieldName => {
      dummyTest.passing(fieldName);
    });
  })();
}

export function untested(fields?: OneOrMoreOf<string>) {
  const suite = createSuite(fields, fieldName => {
    dummyTest.failing(fieldName);
  });
  return suite.get();
}

function createSuiteRunResult(
  fieldNames: Maybe<string[] | string>,
  callback: (fieldName?: string) => void // eslint-disable-line no-unused-vars
) {
  return createSuite(fieldNames, callback)();
}

function createSuite(
  fieldNames: Maybe<string[] | string> = 'field_1',
  callback: (fieldName?: string) => void // eslint-disable-line no-unused-vars
) {
  return create(() => {
    asArray(fieldNames).forEach(fieldName => callback(fieldName));
  });
}

export function ser<F extends TFieldName, G extends TGroupName>(
  res: SuiteResult<F, G>
) {
  return JSON.parse(JSON.stringify(res));
}
