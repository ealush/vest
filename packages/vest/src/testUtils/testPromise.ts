export function TestPromise(cb: (done: () => void) => void): Promise<void> {
  return new Promise<void>(done => cb(done));
}
