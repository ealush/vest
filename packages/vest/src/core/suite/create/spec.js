import faker from 'faker';
import { noop } from 'lodash';
import { dummyTest } from '../../../../testUtils/testDummy';
import create from '.';

describe('Test createSuite module', () => {
  describe.skip('Test Arguments');

  describe('Return value', () => {
    it('should be a function', () => {
      expect(typeof create('suiteName', noop)).toBe('function');
    });

    test('returned function name is `validate`', () => {
      expect(create('boop', noop).name).toBe('validate');
    });
  });

  describe('When returned function is invoked', () => {
    it('Calls `tests` argument', () =>
      new Promise(done => {
        const validate = create('FormName', done);
        validate();
      }));

    it('Passes all arguments over to tests callback', () => {
      const testsCallback = jest.fn();
      const params = [
        1,
        2,
        3,
        { [faker.random.word()]: [1, 2, 3] },
        false,
        [faker.random.word()],
      ];
      const validate = create('FormName', testsCallback);
      validate(...params);
      expect(testsCallback).toHaveBeenCalledWith(...params);
    });
  });

  describe('Initial run', () => {
    const testsCb = jest.fn();
    const genValidate = () => create('initial_run_spec', testsCb);

    it('Should initialize with an empty result object', () => {
      const validate = genValidate();
      expect(Object.keys(validate.get().tests)).toHaveLength(0);
      expect(Object.keys(validate.get().groups)).toHaveLength(0);

      expect(validate.get().errorCount).toBe(0);
      expect(validate.get().warnCount).toBe(0);
      expect(validate.get().testCount).toBe(0);

      expect(validate.get()).toMatchSnapshot();
    });

    it('Should be able to get the suite from the result of createSuite', () => {
      const testsCb = jest.fn();
      expect(create('test_get_suite', testsCb).get()).toMatchSnapshot();
    });

    it('Should be able to reset the suite from the result of createSuite', () => {
      const testSuite = create('test_reset_suite', () => {
        dummyTest.failing('f1', 'm1');
      });
      testSuite();
      expect(testSuite.get().hasErrors()).toBe(true);
      expect(testSuite.get().testCount).toBe(1);
      testSuite.reset();
      expect(testSuite.get().hasErrors()).toBe(false);
      expect(testSuite.get().testCount).toBe(0);
    });

    it('Should return without calling tests callback', () => {
      const validate = create(testsCb);
      expect(testsCb).not.toHaveBeenCalled();
      validate();
      expect(testsCb).toHaveBeenCalled();
    });
  });
});
