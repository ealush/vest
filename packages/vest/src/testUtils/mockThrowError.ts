import { TVestMock } from './TVestMock';

export default function mockThrowError() {
  const deferThrow = jest.fn<string, any>();
  jest.resetModules();
  jest.mock('vest-utils', () => ({
    ...jest.requireActual('vest-utils'),
    deferThrow,
  }));
  const vest = require('vest') as TVestMock;

  return {
    deferThrow,
    vest,
  };
}
