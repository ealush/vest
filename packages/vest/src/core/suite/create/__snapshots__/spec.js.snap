// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test createSuite module Initial run Should initialize with an empty state object 1`] = `
Array [
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": Array [],
    "name": "initial_run_spec",
    "pending": Array [],
    "suiteId": "initial_run_spec",
    "testObjects": Array [],
    "tests": Object {},
  },
  undefined,
]
`;
